sentrysql/conn.go

```diff
@@ -0,0 +1,247 @@
+package sentrysql
+
+import (
+	"context"
+	"database/sql/driver"
+
+	"github.com/getsentry/sentry-go"
+)
+
+type sentryConn struct {
+	originalConn driver.Conn
+	ctx          context.Context
+	config       *sentrySQLConfig
+}
+
+func (s *sentryConn) Prepare(query string) (driver.Stmt, error) {
+	stmt, err := s.originalConn.Prepare(query)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryStmt{
+		originalStmt: stmt,
+		query:        query,
+		ctx:          s.ctx,
+		config:       s.config,
+	}, nil
+}
+
+func (s *sentryConn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	// should only be executed if the original driver implements ConnPrepareContext
+	connPrepareContext, ok := s.originalConn.(driver.ConnPrepareContext)
+	if !ok {
+		return nil, driver.ErrSkip
+	}
+
+	stmt, err := connPrepareContext.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryStmt{
+		originalStmt: stmt,
+		query:        query,
+		ctx:          ctx,
+		config:       s.config,
+	}, nil
+}
+
+func (s *sentryConn) Close() error {
+	return s.originalConn.Close()
+}
+
+func (s *sentryConn) Begin() (driver.Tx, error) {
+	tx, err := s.originalConn.Begin() //nolint:staticcheck We must support legacy clients
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryTx{originalTx: tx, ctx: s.ctx, config: s.config}, nil
+}
+
+func (s *sentryConn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx, error) {
+	// should only be executed if the original driver implements ConnBeginTx
+	connBeginTx, ok := s.originalConn.(driver.ConnBeginTx)
+	if !ok {
+		// fallback to the so-called deprecated "Begin" method
+		return s.Begin()
+	}
+
+	tx, err := connBeginTx.BeginTx(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryTx{originalTx: tx, ctx: s.ctx, config: s.config}, nil
+}
+
+//nolint:dupl
+func (s *sentryConn) Query(query string, args []driver.Value) (driver.Rows, error) {
+	// should only be executed if the original driver implements Queryer
+	queryer, ok := s.originalConn.(driver.Queryer) //nolint:staticcheck We must support legacy clients
+	if !ok {
+		return nil, driver.ErrSkip
+	}
+
+	parentSpan := sentry.SpanFromContext(s.ctx)
+	if parentSpan == nil {
+		return queryer.Query(query, args)
+	}
+
+	span := parentSpan.StartChild("db.sql.query", sentry.WithDescription(query))
+	if s.config.databaseSystem != "" {
```

Comment:
Instead of repeating this for times, perhaps implement a method on `sentrySQLConfig` to set the Span Data?